hooks/bootstrap.py

```diff
@@ -0,0 +1,154 @@
+
+        The ``kwargs`` is there for forward compatibility with future versions of Toolkit.
+
+        :param shotgun: Connection to the Shotgun site.
+        :type shotgun: :class:`shotgun_api3.Shotgun`
+
+        :param pipeline_configuration_id: Id of the pipeline configuration we're bootstrapping into.
+            If None, the ToolkitManager is bootstrapping into the base configuration.
+        :type pipeline_configuration_id: int
+
+        :param configuration_descriptor: Configuration the manager is bootstrapping into.
+        :type configuration_descriptor: :class:`sgtk.descriptor.ConfigDescriptor`
+        """
+        self.shotgun = shotgun
+        self.pipeline_configuration_id = pipeline_configuration_id
+        self.configuration_descriptor = configuration_descriptor
+
+    def can_cache_bundle(self, descriptor):
+        """
+        Indicates if a bundle can be cached by the ``populate_bundle_cache_entry`` method.
+
+        This method is invoked when the bootstrap manager wants to a bundle used by a configuration.
+
+        :param descriptor: Descriptor of the bundle that needs to be cached.
+        :type descriptor: :class:`~sgtk.descriptor.Descriptor`
+
+        :returns: ``True`` if the bundle cache be cached, ``False`` if not.
+        :rtype: bool
+        """
+        return False
+
+    def populate_bundle_cache_entry(self, destination, descriptor, **kwargs):
+        """
+        Populates an entry from the bundle cache.
+
+        This method will be invoked for every bundle for which ``can_cache_bundle``
+        returned ``True``. The hook is responsible for writing the bundle inside
+        the destination folder. If an exception is raised by this method, the files
+        will be deleted from disk and the bundle cache will be left intact.
+
+        Be careful to properly copy all the files or the cache for this bundle
+        will be left in an inconsistent state.
+
+        :param str destination: Folder where the bundle needs to be written. Note
+            that this is not the final destination folder inside the bundle cache.
+
+        :param descriptor: Descriptor of the bundle that needs to be cached.
+        :type descriptor: :class:`~sgtk.descriptor.Descriptor`
+        """
+        raise NotImplementedError("BootstrapHook.populate_bundle_cache_entry is not implemented.")
+
+# The following is an example of how you can use this hook to download
+# applications that are uploaded to a custom entity in Shotgun. Copy this
+# file into your core/hooks folder and
+#
+# class BootstrapHook(get_hook_baseclass()):
+#     """
+#     This is an example bootstrap hook. If allows to download certain bundles from
+#     Shotgun instead of official source for the bundle.
+#     """
+#
+#     # Change this entity name to one that is unused on your Shotgun site. This custom
+#     # non project entity type will be used to upload bundles to Shotgun.
+#     CUSTOM_ENTITY  = "CustomNonProjectEntity01"
```

Comment:
i don't think we need to double up - can we just put the commented out code straight into the hook code above?